src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs

```diff
@@ -0,0 +1,850 @@
+                                            ChannelBinding channelBinding,
+                                            ProtectionLevel protectionLevel,
+                                            TokenImpersonationLevel impersonationLevel
+                                            )
+        {
+            if (_exception != null && !_canRetryAuthentication)
+            {
+                throw _exception;
+            }
+
+            if (_context != null && _context.IsValidContext)
+            {
+                throw new InvalidOperationException(SR.net_auth_reauth);
+            }
+
+            if (credential == null)
+            {
+                throw new ArgumentNullException("credential");
+            }
+
+            if (servicePrincipalName == null)
+            {
+                throw new ArgumentNullException("servicePrincipalName");
+            }
+
+            if (impersonationLevel != TokenImpersonationLevel.Identification &&
+                impersonationLevel != TokenImpersonationLevel.Impersonation &&
+                impersonationLevel != TokenImpersonationLevel.Delegation)
+            {
+                throw new ArgumentOutOfRangeException("impersonationLevel", impersonationLevel.ToString(), SR.net_auth_supported_impl_levels);
+            }
+
+            if (_context != null && IsServer != isServer)
+            {
+                throw new InvalidOperationException(SR.net_auth_client_server);
+            }
+
+            _exception = null;
+            _RemoteOk = false;
+            _framer = new StreamFramer(_innerStream);
+            _framer.WriteHeader.MessageId = FrameHeader.HandshakeId;
+
+            _expectedProtectionLevel = protectionLevel;
+            _expectedImpersonationLevel = isServer ? impersonationLevel : TokenImpersonationLevel.None;
+            _writeSequenceNumber = 0;
+            _readSequenceNumber = 0;
+
+            Interop.SspiCli.ContextFlags flags = Interop.SspiCli.ContextFlags.Connection;
+
+            // A workaround for the client when talking to Win9x on the server side.
+            if (protectionLevel == ProtectionLevel.None && !isServer)
+            {
+                package = NegotiationInfoClass.NTLM;
+            }
+
+            else if (protectionLevel == ProtectionLevel.EncryptAndSign)
+            {
+                flags |= Interop.SspiCli.ContextFlags.Confidentiality;
+            }
+            else if (protectionLevel == ProtectionLevel.Sign)
+            {
+                // Assuming user expects NT4 SP4 and above.
+                flags |= Interop.SspiCli.ContextFlags.ReplayDetect | Interop.SspiCli.ContextFlags.SequenceDetect | Interop.SspiCli.ContextFlags.InitIntegrity;
+            }
+
+            if (isServer)
+            {
+                if (_extendedProtectionPolicy.PolicyEnforcement == PolicyEnforcement.WhenSupported) { flags |= Interop.SspiCli.ContextFlags.AllowMissingBindings; }
```

Comment:
Nit: style-wise, I find such single line if statements with a body contained in braces difficult to read.